arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -78,251 +125,113 @@ void arm_core_mpu_disable(void)
-}
+#endif /* CONFIG_USERSPACE */
 
 /**
- * @brief configure MPU regions for the memory partitions of the memory domain
- *
- * @param   mem_domain    memory domain that thread belongs to
+ * @brief configure fixed (static) MPU regions.
  */
-void arm_core_mpu_configure_mem_domain(struct k_mem_domain *mem_domain)
+void arm_core_mpu_configure_static_mpu_regions(const struct k_mem_partition
+	static_regions[], const u8_t regions_num,
+	const u32_t background_area_start, const u32_t background_area_end)
 {
-	u32_t region_index =
-		_get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
-	u32_t num_partitions;
-	struct k_mem_partition *pparts;
-	struct arm_mpu_region region_conf;
+	if (_mpu_configure_static_mpu_regions(static_regions, regions_num,
+		background_area_start, background_area_end) == -EINVAL) {
 
-	if (mem_domain) {
-		LOG_DBG("configure domain: %p", mem_domain);
-		num_partitions = mem_domain->num_partitions;
-		pparts = mem_domain->partitions;
-	} else {
-		LOG_DBG("disable domain partition regions");
-		num_partitions = 0U;
-		pparts = NULL;
-	}
-
-	for (; region_index < _get_num_regions(); region_index++) {
-		if (num_partitions && pparts->size) {
-			LOG_DBG("set region 0x%x 0x%x 0x%x",
-				    region_index, pparts->start, pparts->size);
-			region_conf.base = pparts->start;
-			_get_ram_region_attr_by_conf(&region_conf.attr,
-				&pparts->attr, pparts->start, pparts->size);
-			_region_init(region_index, &region_conf);
-			num_partitions--;
-		} else {
-			_disable_region(region_index);
-		}
-		pparts++;
+		__ASSERT(0, "Configuring %u static MPU regions failed\n",
+			regions_num);
 	}
 }
 
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
 /**
- * @brief configure MPU region for a single memory partition
- *
- * @param   part_index  memory partition index
- * @param   part        memory partition info
+ * @brief mark memory areas for dynamic region configuration
  */
-void arm_core_mpu_configure_mem_partition(u32_t part_index,
-					  struct k_mem_partition *part)
+void arm_core_mpu_mark_areas_for_dynamic_regions(
+	const struct k_mem_partition dyn_region_areas[],
+	const u8_t dyn_region_areas_num)
 {
-	u32_t region_index =
-		_get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
-	struct arm_mpu_region region_conf;
+	if (_mpu_mark_areas_for_dynamic_regions(dyn_region_areas,
+			dyn_region_areas_num) == -EINVAL) {
 
-	LOG_DBG("configure partition index: %u", part_index);
-
-	if (part &&
-		(region_index + part_index < _get_num_regions())) {
-		LOG_DBG("set region 0x%x 0x%x 0x%x",
-			    region_index + part_index, part->start, part->size);
-		_get_ram_region_attr_by_conf(&region_conf.attr,
-			&part->attr, part->start, part->size);
-		region_conf.base = part->start;
-		_region_init(region_index + part_index, &region_conf);
-	} else {
-		_disable_region(region_index + part_index);
+		__ASSERT(0, "Marking %u areas for dynamic regions failed\n",
+			dyn_region_areas_num);
```

Comment:
Sorry, this is another spot I think I missed earlier.  On a development build this will fail synchronously right here, but on a production build it will plod along having disabled the region.  The intent of __ASSERT() is that this else() clause **should never execute because if it did that would be a bug**, so why does it contain code?

Reply:
I suppose this comment can be resolved.